### PR TITLE
Code carbon integration

### PR DESCRIPTION
**Pull request by laserkelvin**
This PR introduces a `CodeCarbonCallback` to the stack of PyTorch Lightning callbacks in `matsciml`, which integrates `codecarbon` usage throughout training, testing, and inference `pl.Trainer` workflows.

### CODE REVIEW

File: examples/callbacks/code-carbon.py
Comment (melo-gonzo) [inline examples/callbacks/code-carbon.py:1]: I get an error with the `PeriodicProperteisTransform` in this example - we should include the `adaptive_cutoff` flag or bump up the radius.

File: matsciml/lightning/callbacks.py
Comment (melo-gonzo) [inline matsciml/lightning/callbacks.py:1]: Can you add the [`on_exception`](https://github.com/Lightning-AI/pytorch-lightning/blob/682d7ef6b4f8ab412e523f161d95d3e88fbe58cf/src/lightning/pytorch/callbacks/callback.py#L228) method as well to capture the `emissions.pt` data if something kills training, or it's hit with a KeyboardInterrupt?

File: matsciml/lightning/callbacks.py [hunk 1]
```diff
@@ -537,3 +538,140 @@ def write_on_epoch_end(
...
+
+            Parameters
+            ----------
+            trainer
+                Instance of ``pl.Trainer``, which is used to access the
+                logger instances as well.
+            """
+            log_dir = self._temp_output_dir
+            # if nothing was specified
+            if not log_dir:
```
Comment (melo-gonzo) [inline matsciml/lightning/callbacks.py:627]: Can these two `if`'s be combo'd into one line?

**Review by melo-gonzo — CHANGES_REQUESTED** [review]
Looks good overall! Will be good data to have. Left a few comments.

File: matsciml/lightning/callbacks.py (hunk 1 above)
Comment (laserkelvin) [reply matsciml/lightning/callbacks.py]: Addressed in b32e680

File: matsciml/lightning/callbacks.py
Comment (laserkelvin) [reply matsciml/lightning/callbacks.py:1]: Addessed in b20ff2b

File: examples/callbacks/code-carbon.py
Comment (laserkelvin) [reply examples/callbacks/code-carbon.py:1]: Fixed in 8c8dc4e

**Comment by laserkelvin** [comment]
Thanks for all the great feedback! The `on_exception` suggestion actually got me thinking to rewrite it in a much better way too!

**Comment by melo-gonzo** [comment]
Thanks for the clean-ups, approved!

**Review by melo-gonzo — APPROVED** [review]

**Pull request merged**